rcl/test/rcl/test_events.cpp

```diff
@@ -0,0 +1,546 @@
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+  // wait for lease duration to expire
+  std::this_thread::sleep_for(LIVELINESS_LEASE_DURATION_IN_S + milliseconds(500));
+
+  // wait for events
+  bool msg_ready, subscription_event_ready, publisher_event_ready;
+  ASSERT_EQ(wait_for_msgs_and_events(&subscription, &subscription_event, nullptr,
+    context_ptr, 1000, msg_ready, subscription_event_ready, publisher_event_ready), RCL_RET_OK);
+
+  // test that the message published to topic is as expected
+  EXPECT_TRUE(msg_ready);
+  {
+    test_msgs__msg__Primitives msg;
+    test_msgs__msg__Primitives__init(&msg);
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+      test_msgs__msg__Primitives__fini(&msg);
+    });
+    ret = rcl_take(&subscription, &msg, nullptr);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+    EXPECT_EQ(std::string(msg.string_value.data, msg.string_value.size), std::string(test_string));
+  }
+
+  // test subscriber/datareader liveliness changed status
+  EXPECT_TRUE(subscription_event_ready);
+  {
+    rmw_liveliness_changed_status_t liveliness_status;
+    ret = rcl_take_event(&subscription_event, &liveliness_status);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+    EXPECT_EQ(liveliness_status.alive_count, 0);
+    int32_t alive_count_changed = 0;
+    if (is_opensplice) {
+      alive_count_changed = 2;
+    }
+    EXPECT_EQ(liveliness_status.alive_count_change, alive_count_changed);
+    EXPECT_EQ(liveliness_status.not_alive_count, 0);
+    EXPECT_EQ(liveliness_status.not_alive_count_change, 0);
+  }
+  // test that the killed publisher/datawriter has no active events
+  EXPECT_FALSE(publisher_event_ready);
+
+  // clean up
+  ret = rcl_event_fini(&subscription_event);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  ret = rcl_subscription_fini(&subscription, this->node_ptr);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+}
+
+/*
+ * Basic test of publisher and subscriber deadline events, with first message sent after deadline
+ */
+TEST_F(CLASSNAME(TestEventFixture, RMW_IMPLEMENTATION), test_pubsub_deadline_missed)
+{
+  if (is_unsupported) {
+    return;
+  }
+  setup_publisher_and_subscriber(RCL_PUBLISHER_OFFERED_DEADLINE_MISSED,
+    RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED);
+  rcl_ret_t ret;
+
+  // publish message to topic
+  const char * test_string = "testing";
+  {
+    test_msgs__msg__Primitives msg;
+    test_msgs__msg__Primitives__init(&msg);
+    ASSERT_TRUE(rosidl_generator_c__String__assign(&msg.string_value, test_string));
+    ret = rcl_publish(&publisher, &msg);
+    test_msgs__msg__Primitives__fini(&msg);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  }
+
+  // wait for events
+  bool msg_ready, subscription_event_ready, publisher_event_ready;
+
+  // wait for lease duration to expire
+  std::this_thread::sleep_for(DEADLINE_PERIOD_IN_S + milliseconds(500));
```

Comment:
Again, any sleeping in a test needs to be justified in my opinion and avoided entirely if possible. They usually lead to flaky tests due to assuming that the timing will behave in CI every time like it does locally on your machine.

There are cases where it makes sense, but I'd rather be waiting for an event (and knowing or expecting to wait at least some period of time blocking for that event) rather than sleeping.

Reply:
In this case, we could modify this test to do 2 wait() function calls, one for message retrieval and one for event condition.
Changing

Reply:
That's sort of what I was thinking as well. If you can't do it that's ok, but please have a comment justifying the sleep if you cannot. Thanks!

Reply:
With the change to 2 waits, I am running into some issues with this particular test case (`test_pubsub_deadline_missed`) for Connext (but not OpenSplice). `rcl_wait()` is returning OK (not timing out), but nothing in the wait_set is ready.

Reply:
This has been fixed by basically looping around `rcl_wait()`.